### PR TITLE
[C++] export symbols for Windows shared library

### PR DESCRIPTION
**Pull request by chris-b1**
The main change here is adding `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` which exports everything on Windows without adding a bunch of ``__declspec(dllexport)`` to the api, described further [here](https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/)

The other tweaks were basically trial and error to get it to find the dependencies and build/test.  I'm not very experienced with cmake, so if something I've changed looks bad/strange, please let me know!

### CODE REVIEW

**Comment by wesm** [comment]
Cool; I figure at some point we may want to do explicit visibility macros but this helps for now

File: cpp/src/feather/CMakeLists.txt [hunk 1]
```diff
@@ -21,6 +21,16 @@ set(LIBFEATHER_SRCS
...
+  api.h
+  buffer.h
+  common.h
+  feather-c.h
+  io.h
+  metadata.h
+  reader.h
+  status.h
+  types.h
+  writer.h
```
Comment (wesm) [inline cpp/src/feather/CMakeLists.txt]: Can you explain this? Is this something required to make `WINDOWS_EXPORT_ALL_SYMBOLS` work?

**Comment by wesm** [comment]
Do you have the inclination to work on an Appveyor setup for Feather?

File: cpp/src/feather/CMakeLists.txt (hunk 1 above)
Comment (chris-b1) [inline cpp/src/feather/CMakeLists.txt]: This isn't strictly necessary, it just makes it so the header files are pulled into the MSVC solution, compiles either way.

**Comment by chris-b1** [comment]
It will be next week, but yes, I can try to get Appveyor working too.

File: cpp/src/feather/CMakeLists.txt (hunk 1 above)
Comment (wesm) [inline cpp/src/feather/CMakeLists.txt]: Hate to get nitpicky, but can you put the headers in a separate list (`set(LIBFEATHER_HEADERS ...`), then you can list `${LIBFEATHER_HEADERS}` after `${LIBFEATHER_SRCS}` which should do the trick. Thanks! =)

**Comment by chris-b1** [comment]
@wesm - it's a bit hacked together, but here's an Appveyor build working.

https://ci.appveyor.com/project/chris-b1/feather/build/1.0.5

**Comment by wesm** [comment]
Sweet, thank you! I will merge this and enable appveyor for the repo

**Pull request closed without merging**